### DISCUSSION TITLE
MFront wrapper for large strain UMAT

### DISCUSSION THREAD

**Discussion by ngrilli** [thelfer/tfel#321] [Q&A]
I am trying to link a legacy large strain UMAT, i.e. based on the deformation gradient, to MFront.

I started from the example: 
MFrontGallery/unit-tests/mfront-wrappers/fortran/umat/v1/SmallStrainUmatWrapper_v1.mfront

and switched from 
@DSL Default;
to
@DSL ImplicitFiniteStrain;

Additionally, I tried to pass the deformation gradient to the umat. See full code below.

I compile using:

mfront-3.4.0 --obuild --interface=aster MFrontUmatWrapper.mfront

and I get:

BehaviourDSLCommon::analyse: error while treating keyword '@TangentOperator'

1) How do I define the tangent operator in ImplicitFiniteStrain?

Moreover, if I comment out the TangentOperator I get:

error: static assertion failed: N != 0
   static_assert(x,#x)
   ^
/opt/salome_meca/Salome-V2021-s9/prerequisites/Mfront-TFEL340/include/TFEL-3.4.0/TFEL/Math/fsarray.hxx:233:7: note: in expansion of macro ‘TFEL_STATIC_ASSERT’
       TFEL_STATIC_ASSERT(N != 0);

Additionally, my umat uses many other input parameters such as:
COORDS
NOEL
NPT
KSTEP
TIME

2) Is it possible to get those variables in the MFront wrapper?

Thank you very much in advance for your help,
Best Regards,
Nicolò

I paste here the full mfront code and the hxx file:

@DSL ImplicitFiniteStrain;
@Behaviour LargeStrainUmatWrapper;
@Author Nicolò Grilli;
@Date 25 / 08 / 2022;
@Description {
A finite strain wrapper for the UMAT,
so that finite strain UMAT can be used with code_aster
}

@ModellingHypothesis Tridimensional;

@MaterialProperty stress E;
E.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");

@Includes {
#include "/home/nicolo/projects/smeca2021/MFrontUmatWrapper/UmatWrapper.hxx"

#ifndef MFRONT_UMAT_FUNCTION_DECLARATION
#define MFRONT_UMAT_FUNCTION_DECLARATION 1

extern "C" {

void umat_(
    AbaqusRealType *const /* STRESS */, /* stress                   */
    AbaqusRealType *const /* STATEV */, /* internal state variables */
    AbaqusRealType *const /* DDSDDE */, /* tangent operator         */
    AbaqusRealType *const /* SSE */,
    AbaqusRealType *const /* SPD */,
    AbaqusRealType *const /* SCD */,
    AbaqusRealType *const /* RPL */,
    AbaqusRealType *const /* DDSDDT */,
    AbaqusRealType *const /* DRPLDE */,
    AbaqusRealType *const /* DRPLDT */,
    const AbaqusRealType *const /* STRAN */,  /* strain tensor    */
    const AbaqusRealType *const /* DSTRAN */, /* strain increment */
    const AbaqusRealType *const /* TIME */,
    const AbaqusRealType *const /* DTIME */,  /* time increment   */
    const AbaqusRealType *const /* TEMP */,   /* temperature      */
    const AbaqusRealType *const /* DTEMP */,  /* temperature increment    */
    const AbaqusRealType *const /* PREDEF */, /* external state variables */
    const AbaqusRealType
        *const /* DPRED */, /* external state variables increments   */
    const char *const /* CMNAME */,
    const AbaqusIntegerType *const /* NDI */,
    const AbaqusIntegerType *const /* NSHR */,
    const AbaqusIntegerType
        *const /* NTENS */, /* number of components of tensors */
    const AbaqusIntegerType
        *const /* NSTATV */, /* number of internal state variables */
    const AbaqusRealType *const /* PROPS */, /* material properties */
    const AbaqusIntegerType
        *const /* NPROPS */, /* number of material properties */
    const AbaqusRealType *const /* COORDS */,
    const AbaqusRealType *const /* DROT, incremental rotation matrix */,
    const AbaqusRealType
        *const /* PNEWDT, estimation of the next time increment */,
    const AbaqusRealType *const /* CELENT */,
    const AbaqusRealType *const /* DFGRD0 */,
    const AbaqusRealType *const /* DFGRD1 */,
    const AbaqusIntegerType *const /* NOEL */,
    const AbaqusIntegerType *const /* NPT */,
    const AbaqusIntegerType *const /* LAYER */,
    const AbaqusIntegerType *const /* KSPT */,
    const AbaqusIntegerType *const /* KSTEP */,
    AbaqusIntegerType *const /* KINC */,
    const int /* hidden fortran parameter */);

} // end of extern "C"

#endif MFRONT_UMAT_FUNCTION_DECLARATION 1

}

@LocalVariable StiffnessTensor K;

// F0 and F1 are the deformation gradient
// at the beginning and at the end of the time step
@Integrator {

  //
  const AbaqusRealType drot[9] = {1, 0, 0,  //
                                  0, 1, 0,  //
                                  0, 0, 1};
  const AbaqusIntegerType KSTEP[3u] = {0, 0, 0};
  // returned value
  AbaqusIntegerType KINC = 1;
  //
  const char name[81] =
      "Elasticity                              "  //
      "                                        ";
  //
  const AbaqusRealType props[2] = {E, nu};
  //
  AbaqusRealType e[6] = {0,0,0,0,0,0};
  AbaqusRealType de[6] = {0,0,0,0,0,0};
  AbaqusRealType s[6];
  AbaqusRealType rdt;
  //
  //mfront_umat_wrapper::convertStrainToAbaqus<hypothesis>(e, &eto[0]);
  //mfront_umat_wrapper::convertStrainToAbaqus<hypothesis>(de, &deto[0]);
  mfront_umat_wrapper::convertStressToAbaqus<hypothesis>(s, &sig[0]);
  //
  const auto nprops = static_cast<AbaqusIntegerType>(2);
  const auto nstatv = static_cast<AbaqusIntegerType>(0);
  const auto ntens = static_cast<AbaqusIntegerType>(6);
  const auto ndi = static_cast<AbaqusIntegerType>(3);
  const auto nshr = static_cast<AbaqusIntegerType>(3);
  const auto noel = static_cast<AbaqusIntegerType>(0);
  const auto npt = static_cast<AbaqusIntegerType>(0);   // ??
  //
  // unused variables
  AbaqusRealType sse, spd, scd, rpl;  // unused variables
  AbaqusRealType ddsddt[6];
  AbaqusRealType drplde[6];
  AbaqusRealType drpldt;
  AbaqusRealType abaqus_time[2];
  AbaqusRealType pred[1];
  AbaqusRealType dpred[1];
  AbaqusRealType isvs[1];
  AbaqusRealType coords[3] = {0, 0, 0};
  
  // Attempt to introduce deformation gradient
  AbaqusRealType dfgrd0[9] = static_cast<AbaqusRealType>(F0);
  AbaqusRealType dfgrd1[9] = static_cast<AbaqusRealType>(F1);
    
  AbaqusRealType celent;
  AbaqusIntegerType layer;
  AbaqusIntegerType kspt;

  umat_(s,        /* stress                   */
        &isvs[0], /* &isvs[0], internal state variables */
        &K(0, 0), /* tangent operator         */
        &sse, &spd, &scd, &rpl, ddsddt, drplde, &drpldt,
        e,                         /* strain tensor    */
        de,                        /* strain increment */
        abaqus_time,                      //
        &dt,                       /* time increment   */
        &T,                        /* temperature      */
        &dT,                       /* temperature increment    */
        pred,                      /* &esvs[0], external state variables */
        dpred,                     /* &desvs[0], external state variables */
        name, &ndi, &nshr, &ntens, /* number of components of tensors */
        &nstatv,                   /* number of internal state variables */
        &props[0],                 /* material properties                   */
        &nprops,                   /* number of material properties */
        coords, drot,              /* rotation matrix                       */
        &rdt,                      /* estimation of the next time increment */
        &celent, dfgrd0, dfgrd1, &noel, &npt, &layer, &kspt, KSTEP, &KINC, 80);
  //
  if(KINC!=1){
    return FAILURE;
  }
  //
  mfront_umat_wrapper::convertStressFromAbaqus<hypothesis>(&sig[0], s);

}

@TangentOperator {
  static_cast<void>(smt);
  mfront_umat_wrapper::convertStiffnessMatrixFromAbaqus<hypothesis>(&Dt(0, 0),
                                                                    &K(0, 0));
}

/*!
 * \file   UmatWrapper.hxx
 * \brief
 * \author Thomas Helfer
 * \date   11/02/2021
 */

#ifndef LIB_MFRONT_UMAT_WRAPPER_HXX
#define LIB_MFRONT_UMAT_WRAPPER_HXX

#include "/home/nicolo/projects/tfel/include/TFEL/Material/ModellingHypothesis.hxx" //"TFEL/Material/ModellingHypothesis.hxx"
#include "/home/nicolo/projects/tfel/include/TFEL/Math/ST2toST2/UmatNormaliseTangentOperator.hxx" //"TFEL/Math/ST2toST2/UmatNormaliseTangentOperator.hxx"

extern "C" {

//! \brief a simple alias
typedef double AbaqusRealType;
//! \brief a simple alias
typedef int AbaqusIntegerType;

}

namespace mfront_umat_wrapper {

  //! a simple alias
  using Hypothesis = tfel::material::ModellingHypothesis::Hypothesis;

  template <Hypothesis H, typename NumType>
  void convertStrainToAbaqus(AbaqusRealType *const e,
                             const NumType *const strain) {
								 
    constexpr auto isHtheHypothesis = (H == Hypothesis::TRIDIMENSIONAL);
		
    if (isHtheHypothesis) {						 
    //if constexpr (H == Hypothesis::TRIDIMENSIONAL) {
      constexpr auto cste = tfel::math::Cste<NumType>::sqrt2;
      e[0] = strain[0];
      e[1] = strain[1];
      e[2] = strain[2];
      e[3] = strain[3] * cste;
      e[4] = strain[4] * cste;
      e[5] = strain[5] * cste;
    } else {
      tfel::raise("Unsupported hypothesis");
    }
  }  // end of convertStrainToAbaqus

  template <Hypothesis H, typename NumType>
  void convertStressToAbaqus(AbaqusRealType *const s, const NumType *const sig) {
	  
    constexpr auto isHtheHypothesis = (H == Hypothesis::TRIDIMENSIONAL);	  
	  
    if (isHtheHypothesis) {
      constexpr auto icste = tfel::math::Cste<NumType>::isqrt2;
      s[0] = sig[0];
      s[1] = sig[1];
      s[2] = sig[2];
      s[3] = sig[3] * icste;
      s[4] = sig[4] * icste;
      s[5] = sig[5] * icste;
    } else {
      tfel::raise("Unsupported hypothesis");
    }
  }  // end of convertStressToAbaqus

  template <Hypothesis H, typename NumType>
  void convertStressFromAbaqus(NumType *const sig, const AbaqusRealType *const s) {
	  
    constexpr auto isHtheHypothesis = (H == Hypothesis::TRIDIMENSIONAL);		  
	  
    if (isHtheHypothesis) {
      constexpr auto cste = tfel::math::Cste<NumType>::sqrt2;
      sig[0] = s[0];
      sig[1] = s[1];
      sig[2] = s[2];
      sig[3] = s[3] * cste;
      sig[4] = s[4] * cste;
      sig[5] = s[5] * cste;
    } else {
      tfel::raise("Unsupported hypothesis");
    }
  }  // end of convertStressFromAbaqus

  template <Hypothesis H, typename NumType>
  void convertStiffnessMatrixFromAbaqus(NumType *const D,
                                        const AbaqusRealType *const K) {
											
    constexpr auto isHtheHypothesis = (H == Hypothesis::TRIDIMENSIONAL);												
											
    if (isHtheHypothesis) {
      tfel::math::UmatNormaliseTangentOperatorBase<3u, NumType>::exe(D, K);
    } else {
      tfel::raise("Unsupported hypothesis");
    }
  }  // end of convertStiffnessMatrixFromAbaqus

}  // end of namespace mfront_umat_wrapper

#endif /* LIB_MFRONT_UMAT_WRAPPER_HXX */

**Comment by thelfer** [comment]
Hi @ngrilli

Interesting topic. 

> How do I define the tangent operator in ImplicitFiniteStrain?

First, I would not use the `ImplicitFiniteStrain` DSL as you deleguate the integration to an external `umat_` routine. I believe that  the `DefaultFiniteStrain` DSL is better.

Then, I would define a local variable named K as follows:

```cxx
@LocalVariable stress K[36];
```

and fill it in the `@Integrator` code block (You can pass `K.data()` to the `umat_`) and then transform the tangent operator in `@TangentOperator<Abaqus>` to match the conventions in `MFront`. 

> Additionally, my umat uses many other input parameters such as:
> COORDS
> NOEL
> NPT
> KSTEP
> TIME
> Is it possible to get those variables in the MFront wrapper?

Not directly.

You shall create external state variables and material properties to hold those information and hope that the calling solver is able to correctly define them. 

For the `TIME`, you can also use an auxiliary state variable as follows:

```cxx
@AuxiliaryStateVariable time t;
....
@UpdateAuxiliaryStateVariables {
t+=dt;
}
```

Best,

P.S. It would be better attach your source files in a zip file. It would much more pratical.

**Comment by ngrilli** [comment]
Hi @thelfer

Following up on this question,
I tried your suggestions and switched to

@DSL DefaultFiniteStrainDSL;

I have also used a variable to store the Jacobian such as:

@LocalVariable stress K[36];

I switched to the tangent operator:

@TangentOperator\<ABAQUS\>

However, I get an error during compilation:

"error: no match for call to ‘(tfel::material::LargeStrainUmatWrapper<(tfel::material::ModellingHypothesis::Hypothesis)6u, double, false>::TangentOperator {aka tfel::material::FiniteStrainBehaviourTangentOperator<3u, double>}) (int, int)’"

I believe this is due to the data type I use for the variable Dt in UmatWrapper.hxx
for the Jacobian output.

I searched for examples of 
@TangentOperator\<ABAQUS\>
in the gallery or in the code
but could not find any.

Another minor thing, the cast operation:

AbaqusRealType dfgrd0 = static_cast<AbaqusRealType>(F0);

to convert the deformation gradient to Abaqus variable
gives an error, I think I can do that with a for loop
but I wonder if there is a smarter way to do such a conversion.

I attach the zip file with the code.

Thank you very much in advance for your help,
Best Regards,
Nicolò

[MFrontUmatWrapper.zip](https://github.com/thelfer/tfel/files/9932038/MFrontUmatWrapper.zip)

**Reply by thelfer** [reply]
Hi @ngrilli,
Please find attached a corrected version with annotations.
Hope this helps !
Thomas

[MFrontUmatWrapper.zip](https://github.com/thelfer/tfel/files/9933061/MFrontUmatWrapper.zip)
